### DISCUSSION TITLE
无法配置 react-query 默认请求

### DISCUSSION THREAD

**Discussion by AmAzing129** [umijs/umi#11295] [Q&A] [answered]
src/app.tsx
```
import { QueryClient, QueryClientProvider } from 'umi';

const API_SERVER = '/path/to/api/server';
export const reactQuery = {
  devtool: { 
    initialIsOpen: true,
  },
  queryClient: {
	  defaultOptions: {
	    queries: {
	      queryFn: async ({ queryKey }) => {
	        const res = await fetch(`${API_SERVER}/${queryKey.join('/')}`);
	        if (res.status !== 200) {
	          throw new Error(res.statusText);
	        }
	        return res.json();
	      }
	    }
	  }
  },

const queryClient = new QueryClient();

export function rootContainer(container: React.ReactNode) {
  return (
    <QueryClientProvider client={queryClient}>{container}</QueryClientProvider>
  );
}
};
```

umirc.ts

```
export default defineConfig({
 plugins: [
    '@umijs/plugins/dist/react-query',
  ],
  reactQuery: {},
});
```

```
// 并没有发起请求，已经显式安装 "@tanstack/react-query": "^4.29.14",
import { useQuery } from 'umi';

const { data: userInfo } = useQuery(['user', 'info']);
```

我哪里漏掉什么了吗？文档也没找到其他信息

**Comment by AmAzing129** [comment]
以及我想将react-query的默认请求和自带的request结合起来使用，该怎么做呢

**Reply by xXAvoraXx** [reply]
你找到這個問題的答案了嗎？ 我想繼續使用 useRequest 但 umijs 使用舊版本的 useRequest 所以我無法緩存查詢。

**Reply by fz6m** [reply]
umi 自带的 request 功能和 react-query 是有冲突的，因为 request 和 react-query 都作为请求的高层封装，比如都有缓存、错误重试等功能，所以更建议自己封装用于底层请求的 axios / fetch ，顶层请求使用 react-query 。

**Reply by xXAvoraXx** [reply]
我有一些不清楚的問題。 umijs/request 使用 axios 進行封裝，並使用 useRequest 創建查詢。 您說它應該在 useQuery 建議中使用，而不是 useRequest，但是 useQuery 可以與 requestErrorConfig 一起使用嗎？ 我必須提出問題，因為沒有足夠的資源來使用 useQuery 而不是 useRequest。 在Umijs文檔中，config.ts文件中僅顯示了定義示例。

**Reply by fz6m** [reply]
request 的 error config 处理本质上是 axios 的响应拦截器，当你使用 react-query 的时候，底层 fetch / axios 是自己封装的，所以自己添加响应拦截器即可，和 request error 处理等效。

**Reply by AmAzing129** [reply]
reactQuery 和 request 使用同一个 axios，这样代码中都可以使用，目前正常工作，供你参考:
![image](https://github.com/umijs/umi/assets/115673583/69768051-2140-4cbe-a19e-d27275c38966)

**Reply by xXAvoraXx** [reply]
Do you have solution for a mapped like below? Although the code below works, typescript cannot detect it.
您有像下面這樣的映射的解決方案嗎？ 儘管下面的代碼可以工作，但打字稿無法檢測到它。

I'm considering continuing to use useQuery with request. Although it uses axios on the basis of request, it has some aspects that make it easier to use.

我正在考慮繼續使用 useQuery 和 request。 雖然它在請求的基礎上使用axios，但它有一些方面使它更容易使用。

```
export const reactQuery: { queryClient: QueryClientConfig } = {
  queryClient: {
    defaultOptions: {
      queries: {
        staleTime: 5 * (60 * 1000), // 5 mins
        cacheTime: 10 * (60 * 1000), // 10 mins
        select: (response: unknown): API.IResult => (response as API.IResult).data,
      },
    },
  },
};
```

**Reply by AmAzing129** [reply]
I did't understand it and can you give more details?

**Reply by xXAvoraXx** [reply]
Response from web api is returning according to below schema. Check out the code blocks below. getCountriesRequest returns type IResult<T> in response. When I use the getCountriesRequest method with useQuery, it directly reflects the return type of the function, so IResult is returned.

What I want is to map the ['data'] prop in IResult directly to the data prop of the useQuery hook. useRequest does this with dataField. With useQuery, I have to use the select method for every query. Although I added the default behavior of the select method in the reactQuery defaultOptions settings, useQuery is not seeing this behavior anywhere in the code development environment. When the code runtime, the select behavior I defined in defaultOptions works, but typescript cannot see this behavior in the code development environment.


 ```
  interface ResultResponse {
     success?: boolean;
     message?: string;
   };
   interface DataResultResponse<T = any> extends ResultResponse {
     data?: T;
   }

   type IResult<T = any> = T extends undefined ? ResultResponse : DataResultResponse<T>;

export async function getCountriesRequest() {
     const msg = await request<API.IResult<DTOS.CountryDto[]>>('/api/constant/GetCountries', {
       method: 'GET',
     });
  
     return msg;
}

   const {
     isInitialLoading: isLoadingCountries,
     data: countries,
     isFetched: isFetchedCountries,
   } = useQuery(['GetCountries'], getCountriesRequest, {
     enabled: true,
     select: (response) => response.data,
   });
```

**Reply by AmAzing129** [reply]
Check this:
```
export const reactQuery = {
  queryClient: {
    defaultOptions: {
      queries: {
        queryFn: async ({ queryKey }: { queryKey: any[] }) => {
          const endpoint = queryKey.reduce(
            (acc, cur) =>
              `${acc}${
                typeof cur === 'object'
                  ? `?${new URLSearchParams(cur)}`
                  : `/${cur}`
              }`,
          );
          const { data } = await axios.get(`/api/${endpoint}`);
          if (data.success) return data.data;
          return Promise.reject(data.message);
        },
      },
    },
  },
};

// getCountriesRequest in your components
const {
    isInitialLoading: isLoadingCountries,
    data: countries,
    isFetched: isFetchedCountries,
  } = useQuery<API.IResult<DTOS.CountryDto[]>>(['constant', 'GetCountries']);

```
You can pass type directly into useQuery, mapping into data.

**Reply by xXAvoraXx** [reply]
While the application is running, it works as you say, but in the development environment, typescript gives an error. For example, see the video below, the response value should not have a data property, the response itself should be DataResultResponse. (DTOS.CountryDto[])

https://github.com/umijs/umi/assets/6988989/19f659c2-2851-4ca0-b0ba-13777f753860

**Reply by AmAzing129** [reply]
type of response is just the T of `useQuery<T>` so it should be 'useQuery<DTOS.CountryDto[]>'
Maybe you can check 'react-query' doc

**Reply by xXAvoraXx** [reply]
Yes I know. I am using useQuery with umijs request. plugin-request handles the errorHandlig mechanism nicely according to the response from the API. That's why I have to use IResult in this part. What I want is to be able to get the data property in IResult with the dataField property as in useRequest, I want to use the same as useQuery.

```
export async function getCountriesRequest() {
    const msg = await request<API.IResult<DTOS.CountryDto[]>>('/api/constant/GetCountries', {
      method: 'GET',
    });
  
    return msg;
}

function useRequest<R extends ResultWithData = any, P extends any[] = any>(
  service: CombineService<R, P>,
  options?: BaseOptions<R['data'], P>,
): BaseResult<R['data'], P>;

```

**Reply by AmAzing129** [reply]
In your case, the function 'getCountriesRequest' is like 'useCountries' based on useQuery rather than comparing with useQuery. They are different. You still need to pass types to request in getCountriesRequest just like you pass types to useQuery.

**Reply by xXAvoraXx** [reply]
unfortunately nothing has changed.

```
  const {
    isInitialLoading: isLoadingCountries,
    data: countries,
    isFetched: isFetchedCountries,
  } = useQuery<API.IResult<DTOS.CountryDto[]>, Error, DTOS.CountryDto[]>(['GetCountries'], getCountriesRequest, {
    enabled: enableGeneralInformationQueries,
  });
```

**Reply by AmAzing129** [reply]
Sry I can't solve this maybe you can ask help for others

**Comment by fz6m** [comment, marked as answer]
request 和 react-query 是不同的请求方案，用 RQ 底层建议自己封装 axios 或者 fetch 。

开启 `reactQuery: {}` 以后就不需要自己加 provider 了，可以在 `app.ts` 里 [导出](https://umijs.org/docs/max/react-query#%E8%BF%90%E8%A1%8C%E6%97%B6%E9%85%8D%E7%BD%AE%E9%A1%B9) 来配置 react query 的 provider ：

```ts
// src/app.ts
export const reactQuery = {
  devtool: { 
    initialIsOpen: true,
  },
  queryClient: {
	  defaultOptions: {
	    queries: {
	      queryFn: async ({ queryKey }) => {
	        const res = await fetch(`${API_SERVER}/${queryKey.join('/')}`);
	        if (res.status !== 200) {
	          throw new Error(res.statusText);
	        }
	        return res.json();
	      }
	    }
	  }
  },
};
```

或者你可以不开启 `reactQuery: {}` ，自己在运行时套一层 react query provider 开启 RQ 。

**Reply by AmAzing129** [reply]
我理解 request 本身就是对 axios 的封装，为什么不能在 RQ 中直接使用，而是再自己从 axois 开始封装呢？或者文档中可以多增加一些 RQ 使用的说明？

**Reply by little-buddy** [reply]
我感觉也是，看了文档都感觉不如直接用 react-query
